### ISSUE TITLE
How to execute a callback after pred = model(image), but before model.loss(gt, pred)?

### ISSUE THREAD

**Issue by StepanVL** [ultralytics/ultralytics#18647]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have a dataset that consists of images, box+class labels and crisp_image masks.
`crisp_image=1` for pixels, where annotators can clearly distinguish background form targets, `crisp_image=0` otherwise.
I can use model.add_callback('on_train_batch_start') to replace the stock batch with my own batch, containing a tupple `(image_crops, crisp_mask_crops, boxes, classes)`. Say, I use `batch_size = 1`. having said that, between 

`pred = model(images)`

and 

`losses = model.loss(gt, pred)`

I need to insert

`area = crisp_mask.shape[0] * crisp_mask.shape[1]`
`res_boxes = []`
`res_labels = []`
`for box, label in (boxes, labels):`
`    x_center, y_center = box[0], box[1]`
`    if  np.sum(crisp_mask[x1:x2, y1:y2]) > 0.5 * area: # or if mask[y_center, x_center] == 1`
`        res_boxes.append(box)`
`        res_labels.append(label)`

That is "only account for predicted boxes that are located on a crisp part of the image".

There seem to be no callback for that.

How to implement this behavior?

### Additional

I've considered painting blurry areas of the image black (or white) like

`for c in range(img.shape[2]):`
`    img[:, :, c] *= crisp_mask`

But this approach caused a lot of issues in my segmentation project.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @StepanVL, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

It looks like you are looking to customize the model behavior between prediction and loss computation. While there doesn’t appear to be a built-in callback for this specific operation, we recommend sharing details of your implementation setup to better assist. If this is a custom training ❓ Question or 🐛 Bug Report, please provide as much information as possible, including:
1. A [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) that demonstrates your implementation and issue.
2. Dataset image examples, crisp mask samples, and any relevant training logs for additional context.
3. Verification that you are following the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To ensure you’re using the latest features and fixes, please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community and Support

Join the Ultralytics community where it suits you best:
- For real-time chat and support, join our [Discord](https://discord.com/invite/ultralytics) 🎧.
- For longer, in-depth discussions, visit [Discourse](https://community.ultralytics.com).
- Engage with other users and share ideas on the [Subreddit](https://reddit.com/r/Ultralytics).

Additionally, one of our engineers will review this issue and assist you further soon. Stay tuned 😊!

**Comment by Y-T-G** [comment]
You can modify the source code
